pkg/controller/gkenetworkparamset/gnpcontroller_validations.go

```diff
@@ -0,0 +1,246 @@
+	isSecondaryRangeUnspecified := params.Spec.PodIPv4Ranges == nil || (params.Spec.PodIPv4Ranges != nil && len(params.Spec.PodIPv4Ranges.RangeNames) == 0)
+	isDeviceModeUnspecified := params.Spec.DeviceMode == ""
+	if isSecondaryRangeUnspecified && isDeviceModeUnspecified {
+		return &gnpValidation{
+			IsValid:      false,
+			ErrorReason:  networkv1.SecondaryRangeAndDeviceModeUnspecified,
+			ErrorMessage: fmt.Sprintf("SecondaryRange and DeviceMode are unspecified. One must be specified."),
+		}, nil
+	}
+
+	// Check if secondary range exists
+	if params.Spec.PodIPv4Ranges != nil && params.Spec.DeviceMode == "" {
+		for _, rangeName := range params.Spec.PodIPv4Ranges.RangeNames {
+			found := false
+			for _, sr := range subnet.SecondaryIpRanges {
+				if sr.RangeName == rangeName {
+					found = true
+					break
+				}
+			}
+			if !found {
+				return &gnpValidation{
+					IsValid:      false,
+					ErrorReason:  networkv1.SecondaryRangeNotFound,
+					ErrorMessage: fmt.Sprintf("secondary range: %s not found in subnet: %s", rangeName, params.Spec.VPCSubnet),
+				}, nil
+			}
+		}
+	}
+
+	// Check if deviceMode is specified at the same time as secondary range
+	if params.Spec.DeviceMode != "" && params.Spec.PodIPv4Ranges != nil && len(params.Spec.PodIPv4Ranges.RangeNames) > 0 {
+		return &gnpValidation{
+			IsValid:      false,
+			ErrorReason:  networkv1.DeviceModeCantBeUsedWithSecondaryRange,
+			ErrorMessage: "deviceMode and secondary range can not be specified at the same time",
+		}, nil
+	}
+
+	//if GNP with deviceMode and referencing VPC or Subnet is referenced in any other existing GNP
+	if params.Spec.DeviceMode != "" {
+		gnpList, err := c.networkClientset.NetworkingV1().GKENetworkParamSets().List(ctx, v1.ListOptions{})
+		if err != nil {
+			return nil, err
+		}
+		for _, otherGNP := range gnpList.Items {
+			isDifferentGNP := params.Name != otherGNP.Name
+			isMatchingVPC := params.Spec.VPC == otherGNP.Spec.VPC
+			isMatchingSubnet := params.Spec.VPCSubnet == otherGNP.Spec.VPCSubnet
+			isParamsNewer := params.CreationTimestamp.After(otherGNP.CreationTimestamp.Time)
+
+			if isDifferentGNP && isMatchingVPC && isParamsNewer {
+				return &gnpValidation{
+					IsValid:      false,
+					ErrorReason:  networkv1.DeviceModeVPCAlreadyInUse,
+					ErrorMessage: fmt.Sprintf("GNP with deviceMode can't reference a VPC already in use. VPC '%s' is already in use by '%s'", otherGNP.Spec.VPC, otherGNP.Name),
+				}, nil
+			}
+
+			if isDifferentGNP && isMatchingSubnet && isParamsNewer {
+				return &gnpValidation{
+					IsValid:      false,
+					ErrorReason:  networkv1.DeviceModeSubnetAlreadyInUse,
+					ErrorMessage: fmt.Sprintf("GNP with deviceMode can't reference a subnet already in use. Subnet '%s' is already in use by '%s'", otherGNP.Spec.VPC, otherGNP.Name),
+				}, nil
+			}
+		}
+	}
+
+	//if GNP with deviceMode and The referencing VPC is the default VPC
+	if params.Spec.DeviceMode != "" {
```

Comment:
I moved this up in the validation order so it will trigger first. I think this is valuable to keep separated out since the user may be unaware that the default GNP that was auto created is the duplicate the other check is referencing.